File: src/main/java/com/github/huoguangjin/multihighlight/config/MultiHighlightConfig.java
package com.github.huoguangjin.multihighlight.config;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;

import org.jdom.Element;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Rammer on 06/02/2017.
 */
@State(name = "MultiHighlight", defaultStateAsResource = true, storages = {
        @Storage("MultiHighlight.xml")
})
public class MultiHighlightConfig implements PersistentStateComponent<Element> {

    private static final String COLOR_LIST_TAG = "list";
    private static final String COLOR_TAG = "color";

    private List<NamedTextAttr> namedTextAttrs;

    public static MultiHighlightConfig getInstance() {
        return ServiceManager.getService(MultiHighlightConfig.class);
    }

    public MultiHighlightConfig() {
        namedTextAttrs = new ArrayList<>();
    }

    public List<NamedTextAttr> getNamedTextAttrs() {
        return namedTextAttrs;
    }

    public void setNamedTextAttrs(List<NamedTextAttr> namedTextAttrs) {
        this.namedTextAttrs = namedTextAttrs;
    }

    @Nullable
    @Override
    public Element getState() {
        final Element rootTag = new Element("root");
        writeColors(rootTag);
        return rootTag;
    }

    private void writeColors(Element element) {
        if (namedTextAttrs == null || namedTextAttrs.isEmpty()) {
            return;
        }

        Element colorListTag = new Element(COLOR_LIST_TAG);
        for (NamedTextAttr namedTextAttr : namedTextAttrs) {
            Element colorTag = new Element(COLOR_TAG);
            namedTextAttr.writeExternal(colorTag);
            colorListTag.addContent(colorTag);
        }

        element.addContent(colorListTag);
    }

    @Override
    public void loadState(Element state) {
        readColors(state);
    }

    private void readColors(Element element) {
        if (namedTextAttrs == null) {
            namedTextAttrs = new ArrayList<>();
        }

        final Element colorListTag = element.getChild(COLOR_LIST_TAG);
        if (colorListTag != null) {
            for (Element colorTag : colorListTag.getChildren(COLOR_TAG)) {
                namedTextAttrs.add(new NamedTextAttr(colorTag));
            }
        }
    }

    @Override
    public String toString() {
        return "MultiHighlightConfig{namedTextAttrs=" + namedTextAttrs.toString() + '}';
    }
}

File: src/main/java/com/github/huoguangjin/multihighlight/config/MultiHighlightConfig.kt
package com.github.huoguangjin.multihighlight.config

import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.components.service
import com.intellij.openapi.util.JDOMExternalizerUtil
import com.intellij.util.xmlb.Constants
import org.jdom.Element

@State(
  name = "MultiHighlight",
  defaultStateAsResource = true,
  storages = [Storage("MultiHighlight.xml")]
)
class MultiHighlightConfig : PersistentStateComponent<Element> {

  var matchCase = false
  var matchWord = false

  var namedTextAttrs = listOf<NamedTextAttr>()
    private set

  fun updateTextAttrs(textAttrs: List<NamedTextAttr>) {
    namedTextAttrs = textAttrs

    TextAttributesFactory.update(textAttrs)
  }

  override fun getState() = Element("root").also { rootTag ->
    JDOMExternalizerUtil.writeField(rootTag, OPTION_MATCH_CASE, matchCase.toString())
    JDOMExternalizerUtil.writeField(rootTag, OPTION_MATCH_WORD, matchWord.toString())
    writeColors(namedTextAttrs, rootTag)
  }

  private fun writeColors(textAttrs: List<NamedTextAttr>, element: Element) {
    if (textAttrs.isEmpty()) {
      return
    }

    element.addContent(Element(COLOR_LIST_TAG).also { colorListTag ->
      textAttrs.forEach { textAttr ->
        colorListTag.addContent(Element(COLOR_TAG).also { colorTag ->
          textAttr.writeExternal(colorTag)
        })
      }
    })
  }

  override fun loadState(state: Element) {
    matchCase = JDOMExternalizerUtil.readField(state, OPTION_MATCH_CASE) == true.toString()
    matchWord = JDOMExternalizerUtil.readField(state, OPTION_MATCH_WORD) == true.toString()
    namedTextAttrs = readColors(state)
  }

  private fun readColors(element: Element): List<NamedTextAttr> {
    val colorListTag = element.getChild(COLOR_LIST_TAG) ?: return listOf()

    return colorListTag.getChildren(COLOR_TAG).map { colorTag ->
      NamedTextAttr(colorTag)
    }
  }

  override fun toString(): String = "MultiHighlightConfig(c=$matchCase, w=$matchWord, attrs=$namedTextAttrs)"

  companion object {

    private const val OPTION_MATCH_CASE = "MATCH_CASE"
    private const val OPTION_MATCH_WORD = "MATCH_WORD"
    private const val COLOR_LIST_TAG = Constants.LIST
    private const val COLOR_TAG = "color"

    @JvmStatic
    fun getInstance(): MultiHighlightConfig = service()
  }
}
